rfd/0086-stream-events-api.md

```diff
@@ -0,0 +1,234 @@
+[Teleport Client]: https://github.com/gravitational/teleport/blob/cf205b01a5aa88fd4fcdb499ff9b9c40c4e5c335/api/client/client.go
+
+## Why
+
+The Cloud team wants to start tracking the number of monthly active users (MAU) in order to understand the usage and growth of Teleport Cloud.
+To achieve this, audit events will be anonymized and sent to some time-series database (TBD).
+For efficiency, the plan is to implement this by leveraging DynamoDB streaming capabilities.
+
+### Goals
+
+* Add a new `StreamEvents` API to the Teleport Client
+* Implement `StreamEvents` for the DynamoDB audit log
+
+### Non-Goals
+* Implement `StreamEvents` all for available audit logs
+* Refactor the event-handler plugin to use the new API
+
+## Details
+
+In this section we detail how Teleport can be extended to achieve the above goals.
+
+### `Client.StreamEvents` API
+
+The [Teleport Client] will be extended with a new `StreamEvents` API similar to the `StreamSessionEvents` API added in [teleport#7360].
+
+```go
+func (c *Client) StreamEvents(ctx context.Context, cursor string) (chan events.StreamEvent, chan error)
+
+func (c *Client) StreamSessionEvents(ctx context.Context, sessionID string, startIndex int64) (chan events.AuditEvent, chan error)
+```
+
+`StreamSessionEvents` returns a channel of `events.AuditEvent`s.
+`StreamEvents` returns instead a channel of `events.StreamEvent`s that contain the same `events.AuditEvent` in addition to a stream `Cursor`.
+This stream `Cursor` can be used to to resume streaming events by passing it as an argument to the `StreamEvents` API.
+
+```go
+type StreamEvent struct {
+	// Event is an audit event.
+	Event AuditEvent
+	// Cursor is a stream cursor that can be used to resume the stream.
+	Cursor string
+}
+```
+
+[teleport#7360]: https://github.com/gravitational/teleport/pull/7360
+
+### `StreamEvents` RPC
+
+These two APIs are build on top of server-streaming RPCs with the same name:
+
+```protobuf
+// StreamEventsRequest is a request to start or resume streaming audit events.
+message StreamEventsRequest {
+    // Cursor is an optional stream cursor that can be used to resume the stream.
+    string Cursor = 1;
+}
+
+message StreamEvent {
+	// Event is a typed gRPC formatted audit event.
+	events.OneOf Event = 1;
+	// Cursor is a stream cursor that can be used to resume the stream.
+	string Cursor = 2;
+}
+
+service AuthService {
+	// ...
+
+	// StreamEvents streams audit events.
+	rpc StreamEvents(StreamEventsRequest) returns (stream StreamEvent);
+	// StreamSessionEvents streams audit events from a given session recording.
+	rpc StreamSessionEvents(StreamSessionEventsRequest) returns (stream events.OneOf);
+
+	// ...
+}
+```
+
+Similarly to the Teleport API call, the `StreamSessionEvents` RPC returns a stream of `events.OneOf`s, while `StreamEvents` returns a stream of `StreamEvent`s that contain an `events.OneOf` and a stream `Cursor`.
+
+### `IAuditLog.StreamEvents` API
```

Comment:
`IAuditLog` is a huge mess - I wonder if we can add this API without touching `IAuditLog`. For example, `StreamSessionEvents` is in `IAuditLog`, but the following implementations just return a `NotImplementedError`:
- `events.WriterLog`
- `firestorevents.Log`
- `events.FileLog`
- `dynamoevents.Log`

This suggests that the abstraction here is broken and it doesn't belong in the `IAuditLog` interface.

I would find it really confusing to have one method (`StreamSessionEvents`) that is only implemented for `events.AuditLog`, and a very similarly named method (`StreamEvents`) that is only implemented for dynamo.

Reply:
That's a good point. I was following exactly what we have for `StreamSessionEvents`.

I'll think about what kind of internal API could work here as an alternative. Thanks!

Reply:
c3a97e8d9124530c0b219ab37d4d99020284df0f changes the RFD in the following way:
- removes `StreamEvents` API from the `api` module and `IAuditLog`
- details changes to `shards.go` so that it can stream any DynamoDB table and supports stream resuming